src/Interpreters/JIT/CompileDAG.cpp

```diff
@@ -0,0 +1,173 @@
+                    throw Exception(ErrorCodes::LOGICAL_ERROR,
+                    "Cannot find native value for constant column with type {}",
+                    node.result_type->getName());
+
+                compiled_values[compiled_values_index] = native_value;
+                break;
+            }
+            case CompileType::FUNCTION:
+            {
+                Values temporary_values;
+                temporary_values.reserve(node.arguments.size());
+
+                for (auto argument_index : node.arguments)
+                {
+                    assert(compiled_values[argument_index] != nullptr);
+                    temporary_values.emplace_back(compiled_values[argument_index]);
+                }
+
+                compiled_values[compiled_values_index] = node.function->compile(builder, temporary_values);
+                break;
+            }
+            case CompileType::INPUT:
+            {
+                compiled_values[compiled_values_index] = input_nodes_values[input_nodes_values_index];
+                ++input_nodes_values_index;
+                break;
+            }
+        }
+
+        ++compiled_values_index;
+    }
+
+    return compiled_values.back();
+}
+
+std::string CompileDAG::dump() const
+{
+    std::vector<std::string> dumped_values;
+    dumped_values.resize(nodes.size());
+
+    size_t dag_size = nodes.size();
+    for (size_t i = 0; i < dag_size; ++i)
+    {
+        const auto & node = nodes[i];
+
+        switch (node.type)
+        {
+            case CompileType::CONSTANT:
+            {
+                const auto * column = typeid_cast<const ColumnConst *>(node.column.get());
+                const auto & data = column->getDataColumn();
+
+                dumped_values[i] = applyVisitor(FieldVisitorToString(), data[0]) + " : " + node.result_type->getName();
+                break;
+            }
+            case CompileType::FUNCTION:
+            {
+                std::string function_dump = node.function->getName();
+                function_dump += '(';
+
+                for (auto argument_index : node.arguments)
+                {
+                    function_dump += dumped_values[argument_index];
+                    function_dump += ", ";
+                }
+
+                if (!node.arguments.empty())
+                {
+                    function_dump.pop_back();
+                    function_dump.pop_back();
+                }
+
+                function_dump += ')';
+
+                dumped_values[i] = std::move(function_dump);
+                break;
+            }
+            case CompileType::INPUT:
+            {
+                dumped_values[i] = node.result_type->getName();
+                break;
+            }
+        }
+    }
+
+    return dumped_values.back();
+}
+
+UInt128 CompileDAG::hash() const
+{
+    SipHash hash;
+    for (const auto & node : nodes)
+    {
+        hash.update(node.type);
+        hash.update(node.result_type->getName());
```

Comment:
This is subtly incorrect.
If we don't mix string length into the hash, the hashed data can be ambiguous.

Reply:
Updated.

Reply:
It seems that it will be done automatically.
```
void update(const std::string & x)
{
    update(x.data(), x.length());
}
```

Reply:
@kitaisreal This code is hashing only bytes of the string, but does not mix the length.

Hashing two strings in advance will be ambiguous:

1. `ab`, `c`
2. `a`, `bc`